### DISCUSSION TITLE
Firmware update seems to fail

### DISCUSSION THREAD

**Discussion by torkelmj** [UtilitechAS/amsreader-firmware#660] [Q&A]
Howdy.

My AMS reader presently runs FW v2.2.21. "New version v2.2.23" is announced in the web page header. 

When I click the update link in the header, I get the "Device is updating, please wait" dialog on-screen. After a little while, the device reboots. BUT it remains at FW v2.2.21. 

I've also attempted downloading the new firmware file, storing it locally - and then applying the update thru the "Select firmware file for upgrade" option in the web UI. Same thing happens. 

Relevant config/environment: 
- Unit is a Pow-U+ on esp32s2
- Connected to external USB power source
- One-click update / data collection enabled

What am I missing?

**Comment by ArnieO** [comment]
Hi Torkel, 

Strange situation, but we have seen the same maybe two times earlier (out of the the around 2000 devices currently in use). 
To be frank: We do not know what the reason is, maybe something in your Wifi network (some kind of timeout?)- but that's just a hypothesis. 

Anyway: There is a third option for upgrading - through cable. 
It is a bit more elaborate - but it works. 
The process is explained in chapter 3.2 of the Firmware User Manual: https://amsleser.no/index.php?controller=attachment&id_attachment=53

**Comment by torkelmj** [comment]
Thanks, I'll try that. Provided I understand the procedure correctly, all settings in the AMS reader will be lost and reconfiguration will be required. I'll have to allocate some more time for that project. 

FWIW, I attempted a slightly different approach - with no success:
- reconfigured AMS reader to connect to a (temporary) WiFi-zone provided by an iPad (4G/5G)
- hooked up the PC to the same (temporary) WiFi-zone
- found the IP address of the AMS reader and connected to it via HTTP
- attempted the FW update procedures (link in header and download to/upload from PC)

As before, no error messages whatsoever - just a short wait, reboot and then back to normal operation, still on v2.2.21. 

I'll attempt the USB procedure later this week. 

Cheers,

**Reply by ArnieO** [reply]
Yes, reflashing by cable will overwrite most of the settings and data stored in the device.
But you can easily export all stored data to a file, and restore it back to the device after the upgrade:
-> Info-page / Backup & restore

**Reply by torkelmj** [reply]
Updated via USB, all OK. 

FWIW, the FW manual should be slightly updated:
9.2 item 7 reads "`esptool –port COM4 erase_flash`" whereas it should read "`esptool -–port COM4 erase_flash`".

Also, it seems that when a previously backed-up configuration file is restored, the "Enable price fetch from remote server" setting is ignored and has to be manually re-applied.

Thanks for great support.

**Comment by ArnieO** [comment]
> FWIW, the FW manual should be slightly updated:
> 9.2 item 7 reads "`esptool –port COM4 erase_flash`" whereas it should read "`esptool -–port COM4 erase_flash`".

Thank you, greatly appreciated! Word-file is edited, so that this will be corrected in next issue.

> Also, it seems that when a previously backed-up configuration file is restored, the "Enable price fetch from remote server" setting is 
> ignored and has to be manually re-applied.

@gskjold Can you look into this?

**Comment by tux-** [comment]
For those that have access to a linux machine:

I had the same issue with mine, and I downloaded both the esptool, and the new release, then ran: `./esptool --chip esp32s2 --port /dev/ttyACM0 --baud 460800 --before default_reset --after hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size detect 0x10000 firmware.bin` and device got upgraded to latest version with wifi configuration preserved. The command may vary depending on your device, etc, but correct one can be found in the `flash.sh` file.

**Comment by ksdale** [comment]
I just recently had an issue with update repeatedly failing. I discovered that the reason was that the power bank I used failed to register the low current that is drawn from the Pow-U device, and timed out. In some cases however, the USB C output may work even on low current devices, which was the case with my sons power bank. To be safe use a USB charger adapter.

**Reply by ArnieO** [reply]
Good point, thank you for sharing! 
Our recommendation is to not try to power the Pow-U from USB-C devices with PD (Power delivery). There is no reverse diode in the current design, so the PD device will see that there is a voltage present, coming from the internal supercapacitor. The consequence is that they refuse to deliver. 

A reverse diode is added in next design update (in sale from early December), so we assume this problem will then be eliminated.
